src/rulesets-new/json-api-rules/resource-object-rules.ts

```diff
@@ -0,0 +1,323 @@
+      rulesContext.operation.method === "patch"),
+  rule: (responseAssertions) => {
+    responseAssertions.added(
+      "not include content for 204 status codes",
+      (response) => {
+        if (response.bodies.length !== 0) {
+          throw new RuleError({
+            message: "expected response to not have content",
+          });
+        }
+      },
+    );
+    responseAssertions.changed(
+      "not include content for 204 status codes",
+      (beforeResponse, response) => {
+        if (response.bodies.length !== 0) {
+          throw new RuleError({
+            message: "expected response to not have content",
+          });
+        }
+      },
+    );
+  },
+});
+
+const contentFor2xxStatusCodes = new ResponseRule({
+  name: "content for 2xx status codes",
+  matches: (response) => response.statusCode !== "204",
+  rule: (responseAssertions) => {
+    responseAssertions.added(
+      "include content for 2xx status codes other than 204",
+      (response) => {
+        if (response.bodies.length === 0) {
+          throw new RuleError({
+            message: "expected response to have content",
+          });
+        }
+      },
+    );
+    responseAssertions.changed(
+      "include content for 2xx status codes other than 204",
+      (beforeResponse, response) => {
+        if (response.bodies.length === 0) {
+          throw new RuleError({
+            message: "expected response to have content",
+          });
+        }
+      },
+    );
+  },
+});
+
+const dataProperty = new ResponseBodyRule({
+  name: "include JSON:API data property for 2xx status codes",
+  matches: (responseBody, rulesContext) =>
+    ["200", "201"].includes(responseBody.statusCode) &&
+    ["get", "post"].includes(rulesContext.operation.method) &&
+    responseBody.contentType === "application/vnd.api+json",
+  rule: (responseAssertions) => {
+    responseAssertions.body.added.matches({
+      schema: {
+        properties: {
+          data: {
+            type: Matchers.string,
+          },
+        },
+      },
+    });
+
+    responseAssertions.body.changed.matches({
+      schema: {
+        properties: {
+          data: {
+            type: Matchers.string,
+          },
+        },
+      },
+    });
+  },
+});
+
+const jsonApiProperty = new ResponseBodyRule({
+  name: "include JSON:API type property for 2xx status codes",
+  matches: (responseBody, rulesContext) =>
+    ["200", "201"].includes(responseBody.statusCode) &&
+    ["patch", "delete"].includes(rulesContext.operation.method) &&
+    responseBody.contentType === "application/vnd.api+json",
+  rule: (responseAssertions) => {
+    responseAssertions.body.added.matches({
+      schema: {
+        properties: {
+          jsonapi: {
+            type: Matchers.string,
+          },
+        },
+      },
+    });
+
+    responseAssertions.body.changed.matches({
+      schema: {
+        properties: {
+          jsonapi: {
+            type: Matchers.string,
```

Comment:
Interesting... could I write my own Matcher object?

One that comes to mind is the common LinkProperty type: a oneOf type union.